### PR TITLE
Auto press emulator buttons

### LINKED ISSUE #28: Programatically press emulator buttons
This feature could be very useful to create automatic tests.

The automatic tests could be inserted in travis build to increase robustness

### PR DESCRIPTION
**Pull request by stdevIvr**
Fixes #28

Changes:
- Added support for pressing emulator's buttons programmatically.

Does this change need to mentioned in CHANGELOG.md?
no

Requires testing
yes

### CODE REVIEW

**Comment by olemis** [comment]
@stdevIvr status for your last commit not reported in Travis

**Comment by isaacvr** [comment]
1- I don't have an option to press both buttons since startup.
2- don't know if it enters to the bootloader mode.
3- Why do you need to the emulator get disconnected?? Do you mean closed at the end of the test??

**Comment by olemis** [comment]
> 1- I don't have an option to press both buttons since startup.
> 2- don't know if it enters to the bootloader mode.

@stdevAlDen do we need the ones above ? Maybe will require combined binary .

> 3- Why do you need to the emulator get disconnected?? Do you mean closed at the end of the test??

To simulate the case of real device disconnected in the real world and connected back again .

**Comment by olemis** [comment]
> > 3- Why do you need to the emulator get disconnected?? Do you mean closed at the end of the test??
> 
> To simulate the case of real device disconnected in the real world and connected back again .

Besides read [this comment](https://github.com/skycoin/hardware-wallet-js/issues/12#issuecomment-467191388) by @Senyoret1 .

File: tiny-firmware/buttons.c [hunk 1]
```diff
@@ -20,6 +20,8 @@
 #include "buttons.h"
 
 struct buttonState button;
+int falseButtonPress;
```
Comment (olemis) [inline tiny-firmware/buttons.c]: `simulateButtonPress`

File: tiny-firmware/buttons.c [hunk 2]
```diff
@@ -20,6 +20,8 @@
...
+int falseButtonType;
```
Comment (olemis) [inline tiny-firmware/buttons.c]: `buttonPressType`

File: tiny-firmware/buttons.c [hunk 3]
```diff
@@ -70,5 +72,22 @@ void buttonUpdate()
 		}
 	}
 
+	#if EMULATOR
+		if ( falseButtonPress ) { /// If a fake button press event is detected, override button state
+			if ( falseButtonType == 0 ) { /// Press NO
```
Comment (olemis) [inline tiny-firmware/buttons.c]: Use an enum type instead of numeric constants

File: tiny-firmware/buttons.h [hunk 4]
```diff
@@ -31,6 +31,8 @@ struct buttonState {
 };
 
 extern struct buttonState button;
+extern int falseButtonPress;
```
Comment (olemis) [inline tiny-firmware/buttons.h]: same as before

File: tiny-firmware/firmware/protect.c [hunk 5]
```diff
@@ -106,6 +106,8 @@ bool protectButton(ButtonRequestType type, bool confirm_only)
 
 	usbTiny(0);
 
+	falseButtonPress = 0;
```
Comment (olemis) [inline tiny-firmware/firmware/protect.c]: `false`

File: tiny-firmware/emulator/buttons.c [hunk 6]
```diff
@@ -34,6 +34,15 @@ uint16_t buttonRead(void) {
 	if (scancodes[SDL_SCANCODE_RIGHT]) {
 		state |= BTN_PIN_YES;
 	}
+	if ( falseButtonPress ) {
+		if ( falseButtonType == 0 ) {
```
Comment (olemis) [inline tiny-firmware/emulator/buttons.c]: enum

**Pull request merged**